ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/TransactionsWriter.scala

```diff
@@ -103,23 +134,46 @@ private[dao] object TransactionsWriter extends TransactionsWriter {
       flatTransactionWitnessesBatch.foreach(_.execute())
       complementWitnessesBatch.foreach(_.execute())
 
+      val contractBatches = ContractsTable.prepareBatchInsert(
+        ledgerEffectiveTime = ledgerEffectiveTime,
+        transaction = transaction,
+        divulgedContracts = divulgedContracts,
+      )
+
+      for ((deleted, deleteContractsBatch) <- contractBatches.deletions) {
+        val deleteWitnessesBatch = contractWitnessesTable.prepareBatchDelete(deleted.toSeq)
+        require(deleteWitnessesBatch.nonEmpty, "Illegal: deleting witnesses without a contract")
+        // Delete the witnesses first to respect the foreign key constraint of the underlying storage
+        deleteWitnessesBatch.get.execute()
+        deleteContractsBatch.execute()
+      }
+
+      for ((_, insertContractsBatch) <- contractBatches.insertions) {
+        insertContractsBatch.execute()
+      }
+
+      val notDeleted: ContractId => Boolean =
+        !contractBatches.deletions.fold(Set.empty[ContractId])(_._1).contains(_)
+
+      // Insert the witnesses last to respect the foreign key constraint of the underlying storage.
+      // Compute and insert new witnesses regardless of whether the current transaction adds new
+      // contracts because it may be the case that we are only adding new witnesses to existing
+      // contracts (e.g. via a fetch).
```

Comment:
Quite confused phrasing. It's probably worth re-writing this.

Reply:
But, hey, it's kind of funny to read it out loud. :joy:

Reply:
This was actually ok, I wanted to highlight another comment.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/5541/commits/a8b94bd0b1513143fc460fab81d670ef7a2f96a2